modules/ml/src/main/java/org/apache/ignite/ml/trees/trainers/columnbased/ColumnDecisionTreeTrainer.java

```diff
@@ -0,0 +1,559 @@
+import org.apache.ignite.ml.math.functions.IgniteFunction;
+import org.apache.ignite.ml.math.functions.IgniteSupplier;
+import org.apache.ignite.ml.trees.ContinuousRegionInfo;
+import org.apache.ignite.ml.trees.ContinuousSplitCalculator;
+import org.apache.ignite.ml.trees.models.DecisionTreeModel;
+import org.apache.ignite.ml.trees.nodes.DecisionTreeNode;
+import org.apache.ignite.ml.trees.nodes.Leaf;
+import org.apache.ignite.ml.trees.nodes.SplitNode;
+import org.apache.ignite.ml.trees.trainers.columnbased.caches.ContextCache;
+import org.apache.ignite.ml.trees.trainers.columnbased.caches.FeaturesCache;
+import org.apache.ignite.ml.trees.trainers.columnbased.caches.FeaturesCache.FeatureKey;
+import org.apache.ignite.ml.trees.trainers.columnbased.caches.ProjectionsCache;
+import org.apache.ignite.ml.trees.trainers.columnbased.caches.ProjectionsCache.RegionKey;
+import org.apache.ignite.ml.trees.trainers.columnbased.caches.SplitCache;
+import org.apache.ignite.ml.trees.trainers.columnbased.caches.SplitCache.SplitKey;
+import org.apache.ignite.ml.trees.trainers.columnbased.vectors.FeatureProcessor;
+import org.apache.ignite.ml.trees.trainers.columnbased.vectors.SplitInfo;
+import org.jetbrains.annotations.NotNull;
+
+import static org.apache.ignite.ml.trees.trainers.columnbased.caches.FeaturesCache.getFeatureCacheKey;
+import static org.apache.ignite.ml.trees.trainers.columnbased.caches.ProjectionsCache.COLUMN_DECISION_TREE_TRAINER_PROJECTIONS_CACHE_NAME;
+import static org.apache.ignite.ml.trees.trainers.columnbased.caches.SplitCache.COLUMN_DECISION_TREE_TRAINER_SPLIT_CACHE_NAME;
+
+/**
+ * This trainer stores observations as columns and features as rows.
+ * Ideas from https://github.com/fabuzaid21/yggdrasil are used here.
+ */
+public class ColumnDecisionTreeTrainer<D extends ContinuousRegionInfo> implements
+    Trainer<DecisionTreeModel, ColumnDecisionTreeTrainerInput> {
+    /**
+     * Function used to assign a value to a region.
+     */
+    private final IgniteFunction<DoubleStream, Double> regCalc;
+
+    /**
+     * Function used to calculate impurity in regions used by categorical features.
+     */
+    private final IgniteFunction<ColumnDecisionTreeTrainerInput, ? extends ContinuousSplitCalculator<D>> continuousCalculatorProvider;
+
+    /**
+     * Categorical calculator provider.
+     **/
+    private final IgniteFunction<ColumnDecisionTreeTrainerInput, IgniteFunction<DoubleStream, Double>> categoricalCalculatorProvider;
+
+    /**
+     * Cache used for storing data for training.
+     */
+    private IgniteCache<RegionKey, List<RegionProjection>> prjsCache;
+
+    /**
+     * Minimal information gain.
+     */
+    private static final double MIN_INFO_GAIN = 1E-10;
+
+    /**
+     * Maximal depth of the decision tree.
+     */
+    private final int maxDepth;
+
+    /**
+     * Size of block which is used for storing regions in cache.
+     */
+    private static final int BLOCK_SIZE = 1 << 4;
```

Comment:
is it BLOCK_SIZE based on something? Or it can be configured and influence on performance and moved to configs?